jscomp/others/js_promise.ml

```diff
@@ -0,0 +1,59 @@
+(* Copyright (C) 2015-2016 Bloomberg Finance L.P.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * In addition to the permissions granted to you by the LGPL, you may combine
+ * or link a "work that uses the Library" with a publicly distributed version
+ * of this file to produce a combined library or application, then distribute
+ * that combined work under the terms of your choosing, with no requirement
+ * to comply with the obligations normally placed on you by section 4 of the
+ * LGPL version 3 (or the corresponding section of a later version of the LGPL
+ * should you choose to use a later version).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
+
+(** Specialized bindings to Promise, note for simplicity,  
+    this binding does not track the error type, it treat it as an opaque type
+    {[
+
+    ]}
+*)
+
+type + 'a t 
+type error (* abstract error type *)
+
+
+external make : (('a -> unit [@bs]) -> ('e -> unit [@bs]) -> unit [@bs.uncurry]) -> 'a t = "Promise" [@@bs.new]
+(* [make (fun resolve reject -> .. )] *)
+external resolve : 'a -> 'a t = "Promise.resolve" [@@bs.val]
+external reject : 'any -> 'a t = "Promise.reject" [@@bs.val]
+external all : 'a t array -> 'a array t = "Promise.all" [@@bs.val]
+external all2 : 'a0 t -> 'a1 t -> ('a0 * 'a1) t = "Promise.all" [@@bs.val]
+external all3 : 'a0 t -> 'a1 t -> 'a2 t -> ('a0 * 'a1 * 'a2 ) t = "Promise.all" [@@bs.val]
+external all4 : 'a0 t -> 'a1 t -> 'a2 t  -> 'a3 t -> ('a0 * 'a1 * 'a2 * 'a3 ) t = "Promise.all" [@@bs.val]
+external all5 : 'a0 t -> 'a1 t -> 'a2 t  -> 'a3 t -> 'a4 t ->   ('a0 * 'a1 * 'a2 * 'a3 * 'a4 ) t = "Promise.all" [@@bs.val]
+external all6 : 'a0 t -> 'a1 t -> 'a2 t  -> 'a3 t -> 'a4 t -> 'a5 t ->    ('a0 * 'a1 * 'a2 * 'a3 * 'a4 * 'a5 ) t = "Promise.all" [@@bs.val]
+
+external race : 'a t array -> 'a t = "Promise.race" [@@bs.val]
+
+external then_ : ('a -> 'b t [@bs.uncurry]) -> 'b t = "then" [@@bs.send.pipe: 'a t]
+
+external (>>=) : 'a  t -> ('a -> 'b t [@bs.uncurry]) -> 'b t = "then" [@@bs.send]
```

Comment:
I am really not a fan of `>>=` either, even for something like `then`, instead of this:
```ocaml
blah >>= doThen
```
You can easily enough do this and it is more clear:
```ocaml
blah |> then doThen
```
If you have a `then` function that is.  I'm a huge fan of minimizing operators and only piping between things, piping also means you need to name things, which means that the operations are named, meaning you know what they do at a glance without some magical `>>=` things.